### PR TITLE
Parse ONNX Expand to `broadcast_with_dims` 

### PR DESCRIPTION
**Pull request by CharlieL7**
* Relies on #2629 
* Makes a new MIGX operator `broadcast_with_dims` that broadcasts a shape with the dimensions given as the second input.
* Needed for parsing ONNX `Expand` when the second input is not constant.

### CODE REVIEW

File: src/onnx/parse_expand.cpp [hunk 1]
```diff
@@ -43,11 +43,19 @@ struct parse_expand : op_parser<parse_expand>
...
-        check_arg_empty(arg_s, "Expand: dynamic shape is not supported");
-        std::vector<std::size_t> dims;
-        arg_s.visit([&](auto input) { dims.assign(input.begin(), input.end()); });
-        auto out_lens = compute_broadcasted_lens(in_lens, dims);
-        return info.add_instruction(make_op("multibroadcast", {{"out_lens", out_lens}}), args[0]);
+        if(arg_s.empty())
+        {
+            // variable dims input
+            return info.add_instruction(make_op("broadcast_with_dims"), args[0], args[1]);
+        }
```
Comment (TedThemistokleous) [inline src/onnx/parse_expand.cpp:50]: Add coverage for this code path

**Review by TedThemistokleous — CHANGES_REQUESTED** [review]
Add additional coverage and handle tidy warnings

**Comment by bpickrel** [comment]
What's the reason this was created as a separate op instead of extra functionality for the `multibroadcast` op?

File: test/ref/broadcast_with_dims.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    std::vector<int32_t> input_data{-3, 3};
+    std::vector<int64_t> dims_data{2, 1};
+    migraphx::parameter_map params;
+    params["x"]    = migraphx::argument(input_shape, input_data.data());
+    params["dims"] = migraphx::argument(dims_shape, dims_data.data());
+    auto result    = p.eval(params).back();
+    auto output    = result.get<int32_t>();
+    EXPECT(output.get_shape().lens() == std::vector<std::size_t>{2, 2});
+    EXPECT(output.get_shape().strides() == std::vector<std::size_t>{0, 1});
+    EXPECT(output(0, 0) == -3);
```
Comment (bpickrel) [inline test/ref/broadcast_with_dims.cpp:54]:
```suggestion
    std::vector<int> gold = {-3, 3, -3, 3};
    EXPECT(output.to_vector() == gold);
```

File: test/ref/broadcast_with_dims.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    std::vector<float> input_data{7, 11};
+    std::vector<int64_t> dims_data{3};
+    migraphx::parameter_map params;
+    params["x"]    = migraphx::argument(input_shape, input_data.data());
+    params["dims"] = migraphx::argument(dims_shape, dims_data.data());
+    auto result    = p.eval(params).back();
+    auto output    = result.get<float>();
+    EXPECT(output.get_shape().lens() == std::vector<std::size_t>{2, 3});
+    EXPECT(output.get_shape().strides() == std::vector<std::size_t>{1, 0});
+    EXPECT(migraphx::float_equal(output(0, 0), 7.f));
```
Comment (bpickrel) [inline test/ref/broadcast_with_dims.cpp:80]:
```suggestion
    auto rr = output.to_vector();
    std::vector<float> gold = {7.f, 7.f, 7.f, 11.f, 11.f, 11.f};
    EXPECT(std::equal(rr.begin(), rr.end(), gold.begin(), gold.end(), migraphx::float_equal));
```

File: test/ref/broadcast_with_dims.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    std::vector<int64_t> dims_data{2, 2};
+    migraphx::shape input_static_shape{migraphx::shape::int32_type, {2}};
+    migraphx::parameter_map params;
+    params["x"]    = migraphx::argument(input_static_shape, input_data.data());
+    params["dims"] = migraphx::argument(dims_shape, dims_data.data());
+    auto result    = p.eval(params).back();
+    auto output    = result.get<int32_t>();
+    EXPECT(output.get_shape().lens() == std::vector<std::size_t>{2, 2});
+    EXPECT(output.get_shape().strides() == std::vector<std::size_t>{0, 1});
+    EXPECT(output(0, 0) == -3);
```
Comment (bpickrel) [inline test/ref/broadcast_with_dims.cpp:110]:
```suggestion
    std::vector<int> gold = {-3, 3, -3, 3};
    EXPECT(output.to_vector() == gold);
```

File: test/ref/broadcast_with_dims.cpp (hunk 2 above)
Comment (bpickrel) [reply test/ref/broadcast_with_dims.cpp:54]: This and following code suggestions are supposed to replace multiple EXPECT lines, not just 1...Github interface didn't exactly cooperate.

**Review by bpickrel — CHANGES_REQUESTED** [review]

File: test/op_shape_test.cpp [hunk 5]
```diff
@@ -926,6 +926,30 @@ TEST_CASE(broadcast_for_dot_dyn1)
...
+    using migraphx::shape;
+    shape s0{migraphx::shape::float_type, {2, 4}};
+    shape s1{migraphx::shape::int64_type, {4}};
+    std::size_t max_int = std::numeric_limits<std::size_t>::max();
+    std::vector<shape::dynamic_dimension> dyn_dims(4, shape::dynamic_dimension{0, max_int});
+    expect_shape(
+        shape{shape::float_type, dyn_dims}, migraphx::make_op("broadcast_with_dims"), s0, s1);
+}
+
+TEST_CASE(broadcast_with_dims1)
```
[the remark this answers is not in the document]
Comment (TedThemistokleous) [reply test/op_shape_test.cpp:940]: Extra dim. I think if Charlie adds a similar example to what he's documented in the operator that would cover all bases

**Review by TedThemistokleous — APPROVED** [review]
@CharlieL7 overall looks good. Move so change the one test with your example you mention in the code.  You already have 100% coverage so more just change the test to mirror the docstring.

I've kicked off CI for you on this and the previous runs have been passing/without issue.

**Comment by CharlieL7** [comment]
> What's the reason this was created as a separate op instead of extra functionality for the `multibroadcast` op?

very messy that way, this way is cleaner

File: test/onnx/gen_onnx.py [hunk 6]
```diff
@@ -2203,6 +2203,17 @@ def expand_test():
     return ([shape_const, node], [x], [y])
 
 
+@onnx_test()
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply test/onnx/gen_onnx.py:2206]: no, `Expand` always outputs a dynamic shape.

File: test/onnx/parse/expand_test.cpp [hunk 7]
```diff
@@ -37,3 +37,17 @@ TEST_CASE(expand_test)
     auto prog = optimize_onnx("expand_test.onnx");
     EXPECT(p == prog);
 }
+
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply test/onnx/parse/expand_test.cpp:40]: `broadcast_with_dims` always outputs a dynamic shape

File: test/ref/broadcast_with_dims.cpp (hunk 2 above)
Comment (CharlieL7) [reply test/ref/broadcast_with_dims.cpp:54]: I specifically wanted to test that the output shape had the correct indexing

**Review by TedThemistokleous — APPROVED** [review]

**Pull request merged**